46.md

```diff
@@ -2,161 +2,83 @@ NIP-46
 
-Private keys should be exposed to as few systems - apps, operating systems, devices - as possible as each system adds to the attack surface.
+## Signer Discovery
 
-Entering private keys can also be annoying and requires exposing them to even more systems such as the operating system's clipboard that might be monitored by malicious apps.
+The client must somehow be able to contact the signer through a common relay.
 
+### Started by the signer
 
-## Terms
-
-* **App**: Nostr app on any platform that *requires* to act on behalf of a nostr account.
-* **Signer**: Nostr app that holds the private key of a nostr account and *can sign* on its behalf.
+The remote signer generates a connection token in the form
 
+```
+<npub1...>#<secret>?relay=wss://...&relay=wss://...
+```
 
-## `TL;DR`
+The user copies that token and pastes it in the client UI somehow. Then the client can send events of kind `24133` to the specified relays and wait for responses from the remote signer.
 
+### Started by the client
 
-**App** and **Signer** sends ephemeral encrypted messages to each other using kind `24133`, using a relay of choice. 
+The client generates a QR code in the following form (URL-encoded):
 
-App prompts the Signer to do things such as fetching the public key or signing events.
+```
+nostrconnect://<client-key-hex>?relay=wss://...&metadata={"name":"client-name"}
+```
 
-The `content` field must be an encrypted JSONRPC-ish **request** or **response**.
+The signer scans the QR code and sends a `connect` message to the client in the specified relays.
 
-## Signer Protocol
+## Event payloads
 
-### Messages
+Event payloads are [NIP-04](04.md)-encrypted JSON blobs that look like JSONRPC.
 
-#### Request
+Events sent by the client to the remote signer have the following format:
 
-```json
+```js
 {
-  "id": <random_string>,
-  "method": <one_of_the_methods>,
-  "params": [<anything>, <else>]
+  "pubkey": "<client-key-hex>"
+  "kind": 24133,
+  "tags": [
+    ["p", "<signer-key-hex>"]
+  ],
+  "content": "nip04_encrypted_json({id: <random-string>, method: <see-below>, params: []})",
+  ...
 }
 ```
 
-#### Response
+And the events the remote signer sends to the client have the following format:
 
-```json
-{
-  "id": <request_id>,
-  "result": <anything>,
-  "error": <reason>
-}
+```js
+  "pubkey": "<signer-key-hex>"
+  "kind": 24133,
+  "tags": [
+    ["p", "<client-key-hex>"]
+  ],
+  "content": "nip04_encrypted_json({id: <request-id>, result: <any>, error: <reason-string>})",
+  ...
 ```
 
 ### Methods
 
-
-#### Mandatory 
-
-These are mandatory methods the remote signer app MUST implement:
-
-- **describe**
-  - params []
-  - result `["describe", "get_public_key", "sign_event", "connect", "disconnect", "delegate", ...]`  
+- **connect**
+  - params: [`pubkey`, `secret`]
```

Comment:
```suggestion
  - params: []
```

Reply:
This is changing the API that is already in use out there in a bunch of places.

Reply:
Since `secret` is being introduced can it please be explained what it is and why? @pablof7z can you help here?

I think this is from:

> if the user is not sending a secret to apply a pre-authorized list the connect method is redundant; nsecBunker doesn't depend on it. Ok with me if we make it optional

What is a pre-authorized list?

Reply:
@fiatjaf thats true. @rolznz also has a point that @pablof7z said that 'connect' is not needed for nsecbunker, so why are we adding secret here? if secret is needed here for nsecbunker, can we atleast indicate it is optional for client -> signer initializations?

updated my suggestion, also renamed pubkey to indicate that it is the client pubkey. though maybe this is intended to be the signer pubkey? I actually am not sure but its a good thing to state explicitly.